plugins/org.springframework.ide.eclipse.boot.dash/src-v2/org/springframework/ide/eclipse/boot/dash/cloudfoundry/client/v2/CFRouteBuilder.java

```diff
@@ -0,0 +1,244 @@
+		return this;
+	}
+
+	public CFRouteBuilder host(String host) {
+		this.host = host;
+		setRoute(this.host, this.domain, this.path, this.port);
+		return this;
+	}
+
+	public CFRouteBuilder path(String path) {
+		this.path = path;
+		setRoute(this.host, this.domain, this.path, this.port);
+		return this;
+	}
+
+	public CFRouteBuilder port(int port) {
+		this.port = port;
+		setRoute(this.host, this.domain, this.path, this.port);
+		return this;
+	}
+
+	public CFRouteBuilder from(String desiredUrl, Collection<String> domains) throws Exception {
+		// Based on CLI cf/actors/routes.go and testing CLI directly with different "routes" values:
+		// 1. Paths is not allowed in TCP route (valid TCP route: "tcp.spring.io:8888")
+		// 2. Ports are not allowed in HTTP route (valid HTTP route: "myapps.cfapps.io/pathToApp/home")
+		// 3. Schemes (e.g. "http://") are not allowed in routes values. Anything that has a ":" is assumed to be TCP route followed by a port
+		// 4. Route can just be domain, or host and domain
+		//
+		// Therefore, routes values cannot be treated as URIs or URLs, but a combination of domain, host, path and port
+
+		String matchedDomain = null;
+		String matchedHost = null;
+		String path = findPath(desiredUrl);
+		int port = findPort(desiredUrl);
+		String hostDomain = desiredUrl;
+
+		if (port != CFRoute.NO_PORT && StringUtils.hasText(path)) {
+			throw ExceptionUtil.coreException(
+					"Route: " + desiredUrl + " cannot have both port and path. Port:" + port + " Path:" + path);
+		}
+
+		if (StringUtils.hasText(path) && hostDomain.indexOf(path) > 0) {
+			path(path);
+			hostDomain = hostDomain.substring(0, hostDomain.indexOf(path) - 1);
+		} else if (port != CFRoute.NO_PORT && hostDomain.indexOf(Integer.toString(port)) > 0) {
+			port(port);
+			hostDomain = hostDomain.substring(0, hostDomain.indexOf(Integer.toString(port)) - 1);
+		}
+
+		matchedDomain = findDomain(hostDomain, domains);
+
+		if (!StringUtils.hasText(matchedDomain)) {
+			throw ExceptionUtil.coreException("Unable to parse domain from: " + fullRoute
+					+ ". The domain may not exist in the Cloud Foundry target. Please make sure that the URL uses a valid Cloud domain available in the Cloud Foundry target.");
+		}
+		domain(matchedDomain);
+
+		matchedHost = hostDomain.substring(0, hostDomain.length()-matchedDomain.length());
+		while (matchedHost.endsWith(".")) {
+			matchedHost = matchedHost.substring(0, matchedHost.length()-1);
+		}
+
+		if (StringUtils.hasText(matchedHost)) {
+			host(matchedHost);
+		}
+
+		// Be sure to set the full route:
+		this.fullRoute = desiredUrl;
+		return this;
+	}
+
+	protected String findDomain(String hostDomain, Collection<String> domains) {
+		if (hostDomain == null) {
+			return null;
+		}
+		// find exact match
+		for (String name : domains) {
+			if (hostDomain.equals(name)) {
+				return hostDomain;
+			}
+		}
+		// Otherwise split on the first "." and try again
+		if (hostDomain.indexOf(".") + 1 >= 0) {
```

Comment:
Shouldn't that be just `> 0` rather than `>= 0`? if `hostDomain.indexOf(".")` returns `-1` the check succeeds, but it probably shouldn't, right?